File: frontend/lib/sse.ts
/**
 * The streaming client — §1.1's wire protocol, read in the browser.
 *
 * The four event shapes below were transcribed from the contract in Phase 1 and
 * have not changed since; `app/streaming/sse.py` is their server-side twin, and a
 * JSON diff between the two files is the tell that one of them drifted. Two
 * properties they were designed for, both now load-bearing:
 *
 * 1. `DoneEvent` is structurally the non-streaming response's provenance block,
 *    so `provenance.ts` gained exactly one function (`fromDoneEvent`) and
 *    `ModelIndicator` gained nothing at all.
 * 2. `RestartEvent` carries the client-side contract: on `restart`, clear the
 *    in-progress bubble entirely, swap the indicator, resume appending. Never
 *    splice two attempts together — half an answer from one model welded to a
 *    full answer from another reads as a broken *model*, and someone will spend
 *    a day blaming the provider for a client bug.
 *
 * **`EventSource` cannot do this.** It is GET-only and cannot set an
 * `Authorization` header, so the transport is `fetch` plus a reader plus manual
 * frame parsing. That is also why the parser is a separate exported class: the
 * framing rules are worth testing on their own, without a network or a session.
 */

import { GATEWAY_BASE, NetworkError, authHeaders, gatewayErrorFrom } from "./api";
import type { ChatCompletionRequest, ExtractionTier, ServedBy, UsageOut } from "./types";

export type MetaEvent = {
  attempt: number;
  slot: string;
  provider: string;
  model: string;
  requested_slot: string;
  conversation_id: string;
  message_id: string;
};

export type DeltaEvent = {
  choices: { delta: { content?: string } }[];
};

export type RestartEvent = {
  reason: string;
  failed: { provider: string; model: string };
  next: { slot: string; provider: string; model: string };
  attempt: number;
  discarded_chars: number;
};

export type DoneEvent = {
  served_by: ServedBy;
  requested_slot: string;
  substituted: boolean;
  attempts: number;
  usage: UsageOut & { wasted_tokens_out?: number };
  degraded: boolean;
  /** How the turn's attachments reached the model (Phase 4): the worst tier of
   * `cache` | `native` | `llm` | `local`, or null when it carried none. */
  extraction_tier?: ExtractionTier | null;
  /** D34's streaming twin of the completion response's own field: how many
   *  older turns D4's fitting dropped to build this answer. */
  messages_dropped?: number;
  /** D32's streaming twin: the pin disclosure. Rides on a successful stream —
   *  it is not an error, and must not be rendered as one (trap 6). */
  warning?: string | null;
  status: "ok" | "failed";
  /** Present only on `status: "failed"` with a non-trivial partial buffer. */
  partial_content?: string;
};

export type StreamEvent =
  | { event: "meta"; data: MetaEvent }
  | { event: "delta"; data: DeltaEvent }
  | { event: "restart"; data: RestartEvent }
  | { event: "done"; data: DoneEvent };

export type StreamHandlers = {
  onMeta?: (event: MetaEvent) => void;
  onDelta?: (event: DeltaEvent) => void;
  /** Clear the bubble. Do not diff, do not splice. */
  onRestart?: (event: RestartEvent) => void;
  onDone?: (event: DoneEvent) => void;
};

// --------------------------------------------------------------------------- //
// Framing
// --------------------------------------------------------------------------- //
const EVENT_NAMES = new Set(["meta", "delta", "restart", "done"]);

/**
 * Bytes off the wire → whole events.
 *
 * Stateful across calls on purpose: a chunk boundary lands wherever TCP puts it,
 * which is routinely in the middle of a `data:` line, so the parser holds the
 * incomplete tail until the rest arrives. That case is the one most likely to be
 * "working" in a test that feeds it one tidy string and broken in production.
 *
 * Two things it declines to do, both deliberately:
 *
 * - **A comment line is not an event.** `: heartbeat` frames arrive whenever the
 *   gateway is waiting on a slow provider, and a client that mistook one for a
 *   real event would render an empty delta every fifteen seconds.
 * - **An unrecognised event name is skipped, not thrown.** Same reasoning as the
 *   open end of `ContentBlock` in `types.ts`: a fifth event type added by a later
 *   phase must degrade to "ignored", never to a broken stream in an old client.
 */
export class SseParser {
  private buffer = "";

  /** Feed a decoded chunk; get back whatever events completed with it. */
  push(chunk: string): StreamEvent[] {
    // Normalized so the frame split has one delimiter to look for rather than
    // three. SSE permits CR, LF and CRLF line endings interchangeably.
    this.buffer += chunk.replace(/\r\n?/g, "\n");

    const events: StreamEvent[] = [];
    let boundary = this.buffer.indexOf("\n\n");

    while (boundary !== -1) {
      const frame = this.buffer.slice(0, boundary);
      this.buffer = this.buffer.slice(boundary + 2);

      const event = parseFrame(frame);
      if (event) events.push(event);

      boundary = this.buffer.indexOf("\n\n");
    }

    return events;
  }
}

function parseFrame(frame: string): StreamEvent | null {
  let name = "";
  const data: string[] = [];

  for (const line of frame.split("\n")) {
    if (line === "" || line.startsWith(":")) continue; // heartbeats and blanks

    const colon = line.indexOf(":");
    const field = colon === -1 ? line : line.slice(0, colon);
    // One optional space after the colon is part of the framing, not the value.
    const value = colon === -1 ? "" : line.slice(colon + 1).replace(/^ /, "");

    if (field === "event") name = value;
    else if (field === "data") data.push(value);
  }

  if (!EVENT_NAMES.has(name) || data.length === 0) return null;

  try {
    // Multi-line `data:` rejoins with newlines — the spec's rule, and free
    // insurance against a provider's text ever being framed across lines.
    return { event: name, data: JSON.parse(data.join("\n")) } as StreamEvent;
  } catch {
    return null;
  }
}

// --------------------------------------------------------------------------- //
// The transport
// --------------------------------------------------------------------------- //
/**
 * Stream one turn, dispatching each event as it lands.
 *
 * Resolves when the response body is exhausted — which, usefully, is *after* the
 * gateway's collector has finished persisting: the orchestrator writes the
 * `messages` row after its final `yield`, so the socket does not close until that
 * is done. A caller revalidating its transcript cache on this promise therefore
 * has no race with the write.
 *
 * Rejects with:
 *
 * - `GatewayError` — a pre-stream failure. Not a special case: the gateway
 *   commits to a 200 only once a candidate has produced its first chunk (D13),
 *   so a fully-down provider pool arrives here as the ordinary error envelope,
 *   `request_id` and all. A mid-stream failure is *not* this; it arrives in-band
 *   as `done` with `status: "failed"`, because by then the status line is spent.
 * - `NetworkError` — the gateway was never reached, or the connection died
 *   before `done`. A truncated stream must not resolve as success.
 *
 * An abort is neither: a cancelled turn is a person changing their mind, and it
 * resolves quietly.
 */
export async function openCompletionStream(
  body: Omit<ChatCompletionRequest, "stream">,
  handlers: StreamHandlers,
  signal?: AbortSignal,
): Promise<void> {
  let response: Response;
  try {
    response = await fetch(`${GATEWAY_BASE}/v1/chat/completions`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...(await authHeaders()) },
      body: JSON.stringify({ ...body, stream: true }),
      signal,
    });
  } catch (error) {
    if (isAbort(error)) return;
    throw new NetworkError();
  }

  if (!response.ok) throw await gatewayErrorFrom(response);
  if (!response.body) throw new NetworkError("The gateway returned no response body.");

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  const parser = new SseParser();
  let sawDone = false;

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;

      for (const event of parser.push(decoder.decode(value, { stream: true }))) {
        if (dispatch(event, handlers)) sawDone = true;
      }
    }
  } catch (error) {
    if (isAbort(error)) return;
    throw new NetworkError();
  } finally {
    // Releases the connection. Without it an aborted turn leaves the gateway
    // streaming into a socket nobody is reading, which is exactly the upstream
    // generation the Stop button is meant to end.
    void reader.cancel().catch(() => {});
  }

  if (!sawDone && !signal?.aborted) {
    throw new NetworkError("The connection closed before the model finished.");
  }
}

/** Route one event to its handler. Returns true for the terminal one. */
function dispatch(event: StreamEvent, handlers: StreamHandlers): boolean {
  switch (event.event) {
    case "meta":
      handlers.onMeta?.(event.data);
      return false;
    case "delta":
      handlers.onDelta?.(event.data);
      return false;
    case "restart":
      handlers.onRestart?.(event.data);
      return false;
    case "done":
      handlers.onDone?.(event.data);
      return true;
  }
}

function isAbort(error: unknown): boolean {
  return error instanceof DOMException && error.name === "AbortError";
}

File: app/streaming/sse.py
"""Server-side SSE framing (§1.1) — the mirror image of Step 7.

Step 7 taught the gateway to read a trickle *from* a provider. This module
teaches it to write a trickle *to* its own client, in the wire format browsers
understand natively: ``event: <name>`` followed by ``data: <json>``, one blank
line apart.

The four event shapes below are ``frontend/lib/sse.ts``'s types transcribed
field-for-field — a JSON diff between the two is the tell that one of them
drifted. ``DoneEvent.usage`` deliberately reuses :class:`~app.schemas.chat.UsageOut`
rather than the ad hoc ``tokens_in``/``tokens_out`` names in the contracts
doc's illustrative snippet: the frontend file's own docstring is explicit that
``DoneEvent`` is meant to be structurally the non-streaming response's
provenance block, so a client that already renders one already renders the
other, and a second usage shape would break that for no benefit.

**This module only formats.** It does not open a provider connection, does not
decide when a ``restart`` happens, and does not merge a heartbeat into a real
event stream — Step 9's orchestrator drives it. What matters most here is not
the framing (that part is easy) but the response headers: leave one out and an
nginx-family proxy will happily buffer the whole body and hand it over as one
blob, producing a gateway that passes every test and does not stream in
production.
"""

from __future__ import annotations

from typing import Final, Literal

from fastapi import Request
from pydantic import BaseModel, ConfigDict

from app.providers.types import ExtractionTier
from app.schemas.chat import ServedBy, UsageOut

SSE_MEDIA_TYPE: Final = "text/event-stream"

SSE_HEADERS: Final[dict[str, str]] = {
    "Cache-Control": "no-cache",
    "Connection": "keep-alive",
    # Not optional. Without this an nginx-family proxy in front of the app
    # buffers the entire body and delivers it as one blob — a gateway that
    # passes every test and, on the deployed instance, visibly does not stream.
    "X-Accel-Buffering": "no",
}

HEARTBEAT_INTERVAL_S: Final = 15.0
"""How often to send a keepalive when nothing real is ready yet.

Comfortably under any proxy's or client's own idle-connection timeout. The
case that matters most is the pre-first-token wait (D13): a slow provider can
legitimately take several seconds to produce anything, and with nothing sent a
client has no way to tell the gateway apart from a dead socket.
"""


# --------------------------------------------------------------------------- #
# §1.1's four event payloads
# --------------------------------------------------------------------------- #
class MetaEvent(BaseModel):
    """The first event on every streamed message, sent before the first delta."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    attempt: int
    slot: str
    provider: str
    model: str
    requested_slot: str
    conversation_id: str
    message_id: str


class _Delta(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    content: str


class _Choice(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    delta: _Delta


class DeltaEvent(BaseModel):
    """One incremental piece of text, OpenAI's ``choices[].delta.content`` shape.

    Kept OpenAI-shaped rather than a bare string so a client that already knows
    how to read a chat-completion chunk needs to learn nothing new to read this.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    choices: list[_Choice]

    @classmethod
    def of(cls, content: str) -> DeltaEvent:
        return cls(choices=[_Choice(delta=_Delta(content=content))])


class _FailedCandidate(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    provider: str
    model: str


class _NextCandidate(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    slot: str
    provider: str
    model: str


class RestartEvent(BaseModel):
    """D1's restart: discard the partial, disclose why, name what comes next.

    The client-side contract this event exists to drive: clear the in-progress
    bubble entirely and swap the model indicator. Never splice or diff two
    attempts together.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    reason: str
    failed: _FailedCandidate
    next: _NextCandidate
    attempt: int
    discarded_chars: int

    @classmethod
    def of(
        cls,
        *,
        reason: str,
        failed_provider: str,
        failed_model: str,
        next_slot: str,
        next_provider: str,
        next_model: str,
        attempt: int,
        discarded_chars: int,
    ) -> RestartEvent:
        """Build one without reaching for the two private nested shapes.

        They are private because nothing outside this module should have an
        opinion about how the event nests; the orchestrator has the six flat
        facts and this turns them into the wire shape.
        """
        return cls(
            reason=reason,
            failed=_FailedCandidate(provider=failed_provider, model=failed_model),
            next=_NextCandidate(slot=next_slot, provider=next_provider, model=next_model),
            attempt=attempt,
            discarded_chars=discarded_chars,
        )


class StreamUsage(UsageOut):
    """``UsageOut`` plus the one number only a discarded attempt produces.

    Tokens a failed attempt actually generated were really spent — the free
    tier charged them against RPD/TPM even though the text is thrown away —
    so they are counted separately from the tokens that made it to the client.
    """

    wasted_tokens_out: int = 0


class DoneEvent(BaseModel):
    """The terminal event. Once this is sent, the message is final (D1) — no
    further ``restart`` may follow it."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    served_by: ServedBy
    requested_slot: str
    substituted: bool
    attempts: int
    usage: StreamUsage
    degraded: bool
    extraction_tier: ExtractionTier | None = None
    """How this turn's attachments reached the model (Phase 4, D25) — ``cache``,
    ``native``, ``llm`` or ``local``, and the worst of them when there was more
    than one. ``None`` when the turn carried no attachment.

    ``degraded`` says *whether* the reading was a fallback; this says *what
    kind*, so the indicator can render "read by local OCR" rather than a bare
    warning — the same disclosure discipline ``served_by`` gets."""

    messages_dropped: int = 0
    """D34's streaming twin of :attr:`~app.schemas.chat.ChatCompletionResponse.messages_dropped`.
    Zero on a replayed cache hit (:func:`stream_cached_completion`) — nothing was
    rendered this turn."""

    warning: str | None = None
    """D32's streaming twin of :attr:`~app.schemas.chat.ChatCompletionResponse.warning`.
    ``None`` on a replayed cache hit — a replay never reruns pin resolution."""

    key_pool: Literal["shared", "private"] | None = None
    """D42's streaming twin of :attr:`~app.schemas.chat.ChatCompletionResponse.key_pool`
    (Phase 6 Step 7). ``None`` on a replayed cache hit — nothing was spent
    this turn — and only otherwise when every candidate was skipped on an
    open breaker, the one case with no resolved credential to name."""

    status: Literal["ok", "failed"]
    partial_content: str | None = None
    """Set only when ``status == "failed"`` and the longest discarded buffer is
    non-trivial, so the client can offer "keep partial answer"."""


StreamEventPayload = MetaEvent | DeltaEvent | RestartEvent | DoneEvent

_EVENT_NAMES: Final[dict[type[BaseModel], str]] = {
    MetaEvent: "meta",
    DeltaEvent: "delta",
    RestartEvent: "restart",
    DoneEvent: "done",
}


# --------------------------------------------------------------------------- #
# Framing
# --------------------------------------------------------------------------- #
def format_event(data: StreamEventPayload) -> bytes:
    """One SSE frame: ``event: <name>\\ndata: <json>\\n\\n``.

    The event name is derived from the payload's own type rather than passed
    alongside it, so framing a :class:`DoneEvent` under ``event: delta`` by a
    typo is a type error caught by mypy, not a confused frontend discovering it
    at runtime.
    """
    name = _EVENT_NAMES[type(data)]
    body = data.model_dump_json()
    return f"event: {name}\ndata: {body}\n\n".encode()


def format_heartbeat() -> bytes:
    """A comment line: valid SSE, silently ignored by every client's ``onmessage``.

    Keeps the connection alive through a gap where nothing has happened yet to
    report, without the client mistaking it for a real event.
    """
    return b": heartbeat\n\n"


# --------------------------------------------------------------------------- #
# Client-disconnect detection
# --------------------------------------------------------------------------- #
async def client_disconnected(request: Request) -> bool:
    """Has the client gone away?

    A thin wrapper around Starlette's own check rather than a call spread
    across the streaming path — a disconnect is not a fault (§1.1: "detect it,
    abort upstream, commit tokens, persist nothing further, do not restart"),
    and every caller deciding that the same way starts with asking it the same
    way.
    """
    return await request.is_disconnected()
